File: LeetCode/SlidingWindow/python/longestSubstringNoReps.py

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        mySet = set()
        maxLen = 0
        n = len(s)
        l = 0

        # Check each chars of s using r pointer
        for r in range(n):
            
            # Check if there are duplicates in our set and update our set
            # Here we're shrinking our window by moving the left pointer
            # and removing the duplicate from our set
            while s[r] in mySet:
                mySet.remove(s[l])
                l += 1
            
            # at this point, our set have no duplicates, so add the char to our set
            mySet.add(s[r])

            # at every iteration, update our max length
            maxLen = max(maxLen, r - l + 1)

        return maxLen
        
    @staticmethod
    def testSolution(input) -> None:
        print("Input: s: {}".format(input["s"]))
        print("Expected: {}".format(input["expected"]))
        val = Solution().lengthOfLongestSubstring(input["s"])
        print("Result: {}".format(val))
        print(("PASS" if val == input["expected"] else "FAIL"))
        print("-" * 50)


if __name__ == "__main__":
    input = {"s": "abcabcbb", "expected": 3}
    Solution.testSolution(input)

    input = {"s": "bbbbb", "expected": 1}
    Solution.testSolution(input)

    input = {"s": "pwwkew", "expected": 3}
    Solution.testSolution(input)

File: LeetCode/SlidingWindow/java/longestSubstringNoReps.java
package LeetCode.SlidingWindow.java;

import java.util.HashSet;
import java.util.Set;

record longestSubstringNoRepsRecord(String s, int expected) {
}

public class longestSubstringNoReps {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> set = new HashSet<>(); // we use a set to keep track of duplicates
        int maxLen = 0;
        int n = s.length();
        int l = 0;

        // Check each chars of s using r pointer
        for (int r = 0; r < n; r += 1) {
            // Check if there are duplicates in our set and update our set
            // Here we're shrinking our window by moving the left pointer
            // and removing the duplicate from our set
            while (set.contains(s.charAt(r))) {
                set.remove(s.charAt(l));
                l += 1;
            }

            // at this point, our set have no duplicates, so add the char to our set
            set.add(s.charAt(r));

            // at every iteration, update our max length
            maxLen = Math.max(maxLen, r - l + 1);

        }

        return maxLen;
    }

    public static void main(String[] args) {
        longestSubstringNoRepsRecord input = new longestSubstringNoRepsRecord("abcabcbb", 3);
        longestSubstringNoReps.testSolution(input);

        input = new longestSubstringNoRepsRecord("bbbbb", 1);
        longestSubstringNoReps.testSolution(input);

        input = new longestSubstringNoRepsRecord("pwwkew", 3);
        longestSubstringNoReps.testSolution(input);
    }

    private static void testSolution(longestSubstringNoRepsRecord input) {
        System.out.println("Input: s: " + input.s());
        System.out.println("Expected: " + input.expected());
        int val = new longestSubstringNoReps().lengthOfLongestSubstring(input.s());
        System.out.println("Result: " + val);
        System.out.println((val == input.expected() ? "PASS" : "FAIL"));
        System.out.println("-".repeat(50));
    }
}
